sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala

```diff
@@ -283,7 +543,32 @@ object Mask {
         transformChar(_, maskUpper, maskLower, maskDigit, maskOther).toChar
       }
     }
-    org.apache.spark.unsafe.types.UTF8String.fromString(transformedString)
+    UTF8String.fromString(transformedString)
+  }
+
+  def mask_first_n(
+      input: Any,
+      charCount: Int,
+      maskUpper: Any,
+      maskLower: Any,
+      maskDigit: Any,
+      maskOther: Any): UTF8String = {
+
+    val transformedString = if (input == null) {
+      null
+    } else {
+      val inputStr = input.asInstanceOf[UTF8String]
+      val stringSize = inputStr.numChars
+      val endIdx = if (stringSize < charCount) stringSize else charCount
+      inputStr.toString.zipWithIndex.map { case (ch, i) =>
```

Comment:
`zipWithIndex` is slow, and this expression will be evaluating for every input row. Can we instead split the string based on `endIdx` and then just call `transformChar` on every char of the first token?

Reply:
Replaced `zipWithIndex` with `for i`